README.md
roundhouse: An R wrapper to the ICNDb API <img src="tools/roundhouse-logo.png" align="right" width="130" height="150" />
========================================================================================================================

<!-- badges: start -->
[![CRAN
status](https://www.r-pkg.org/badges/version/roundhouse)](https://CRAN.R-project.org/package=roundhouse)
[![Travis-CI Build
Status](https://travis-ci.org/bgreenwell/roundhouse.svg?branch=master)](https://travis-ci.org/bgreenwell/roundhouse)
[![AppVeyor Build
Status](https://ci.appveyor.com/api/projects/status/github/bgreenwell/roundhouse?branch=master&svg=true)](https://ci.appveyor.com/project/bgreenwell/roundhouse)
[![Coverage
Status](https://img.shields.io/codecov/c/github/bgreenwell/roundhouse/master.svg)](https://codecov.io/github/bgreenwell/roundhouse?branch=master)
[![lifecycle](https://img.shields.io/badge/lifecycle-maturing-brightgreen.svg)](https://www.tidyverse.org/lifecycle/#stable)
<!-- badges: end -->

Overview
--------

Generate random Chuck Norris facts from the excellent internet Chuck
Norris database 😎.

Sample fact
-----------

Chuck Norris solved the [traveling salesman
problem](https://en.wikipedia.org/wiki/Travelling_salesman_problem) in
*O*(1) time. Here’s the pseudo-code:

    Break salesman into N pieces.
    Kick each piece to a different city.

Installation
------------

You can install the development version of `roundhouse` from GitHub
using

``` r
if (!requireNamespace("devtools")) {
  install.packages("devtools")
}
devtools::install_github("bgreenwell/roundhouse")
```

Example usage
-------------

⚠️ **WARNING:** Some facts may contain explicit material!

``` r
# Generate a random fact
roundhouse::random_fact()
#> [1] "Chuck Norris has never been accused of murder because his roundhouse kicks are recognized as 'acts of God.'"

# Generate multiple random facts
roundhouse::random_fact(3)
#> [1] "Chuck Norris doesn't need garbage collection because he doesn't call .Dispose(), he calls .DropKick()."
#> [2] "There are no steroids in baseball. Just players Chuck Norris has breathed on."                         
#> [3] "All browsers support the hex definitions #chuck and #norris for the colors black and blue."

# Print fact categories
roundhouse::fetch_categories()
#> [1] "explicit" "nerdy"

# Generate a random fact with a roundhouse kick!
roundhouse::kick(width = 40, size = 15)
```

![](tools/roundhouse.gif)

``` r
# Or simply supply your own text!
roundhouse::kick("Chuck Norris can’t test for equality because he has no equal.",
                 type = 2, width = 40, size = 25, fps = 5)
```

<img src="tools/README-example-02-1.gif" style="display: block; margin: auto auto auto 0;" />

Inspirations
------------

-   [Lone Wolf McQuade](https://www.youtube.com/watch?v=pfLTbzU0FXo)


NEWS.md
# roundhouse 0.0.1.9000


## New functions

* You can suggest new Chuck Norris facts be added in the next release via the new `suggest_new_fact()` function!

* New function `punch()` which works just like `kick()` (but without the `type` argument).

## Minor changes

* Generating Chuck Norris facts no longer requires a network connection.

* Fix URL typo in `DESCRIPTION`.

* New `sound` and `sound_repeat` arguments to `random_fact()` for determining whether or not the phrase "Chuck Norris" should be shouted, and how many times, before generating a random fact. You can also delay the sound a certain number of seconds using the `sound_delay` argument.

* The full database of existing facts is now available as a data frame via `roundhouse::facts`.


# roundhouse 0.0.1

* Initial release.


tests/testthat/test-all.R
context("random facts")

test_that("random_fact() works", {
  fact <- random_fact(n = 1L, include = NULL, exclude = NULL)
  expect_is(fact, class = "character")
})


context("roundhouse kicks")

test_that("kick() works", {
  skip_if_not_installed("magick")
  gif1 <- kick()
  gif2 <- kick("Chuck Norris can’t test for equality because he has no equal.",
               type = 2, width = 40, size = 25, fps = 5)
  expect_is(gif1, class = "magick-image")
  expect_is(gif2, class = "magick-image")
  expect_error(kick(type = 3))
})


R/kick.R
#' Roundhouse kick
#'
#' Generate a roundhouse kick from Chuck Norris...with a random fact (or user-
#' supplied text).
#'
#' @param text Character string specifying the text to display with the
#' roundhouse (preferably, a Chuck Norris fact). If \code{NULL}, a fact is
#' selected at random from the ICNDb.
#'
#' @param type Integer specifying which roundhouse kick to generate. Current
#' options are 1 (standard roundhouse kick) or 2 (roundhouse kick bad guy
#' through bar window). Default is 1.
#'
#' @param width Integer specifying the number of characters per line. Default is
#' 35
#'
#' @param size Integer specifying the font size in pixels. Default is 10.
#'
#' @param color Character string specifying the color of the text.
#'
#' @param fps Integer specifying the frames per second of the roundhouse kick.
#' (Must be a factor of 100.)
#'
#' @param loop Integer specifying the number of times the roundhouse kick should
#' be repeated. Though a single roundhouse kick is always sufficient, the
#' default is \code{0} which results in an infinite number of roundhouse kicks.
#'
#' @param ... Additional optional arguments to be passed onto
#' \code{\link[magick]{image_annotate}}.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' # Random fact
#' kick(width = 40, size = 15)
#'
#' # User-supplied text
#' kick("Chuck Norris can’t test for equality because he has no equal.",
#'      type = 2, width = 40, size = 25, fps = 5)
#' }
kick <- function(text = NULL, type = 1, width = 35, size = 20, color = "white",
                 fps = NULL, loop = 0, ...) {
  if (!requireNamespace("magick", quietly = TRUE)) {
    stop("The magick package is required for roundhouse kicks.", call. = FALSE)
  }
  if (!(type %in% 1:2)) {
    stop("Argument `type` must be 1 or 2.")
  }

  if (is.null(text)) {
    text <- random_fact(1L)
  }
  url <- c(
    "https://media.giphy.com/media/l1J3nY7N7LBrBobVm/giphy.gif",
    "https://media.giphy.com/media/26n6MnzMOIrDGQJmo/giphy.gif"
  )[type]
  # path <- system.file("extdata", paste0("kick-0", type, ".gif"),
  #                     package = "roundhouse")
  text <- paste0(strwrap(text, width = width), collapse = "\n")
  gif <- magick::image_read(path = url)
  if (is.null(fps)) {
    magick::image_annotate(
      image = gif, text = text, color = color, size = size, ...
    )
  } else {
    magick::image_animate(
      image = magick::image_annotate(
        image = gif, text = text, color = color, size = size, ...
      ),
      fps = fps,
      dispose = "none",
      loop = loop
    )
  }
}


R/roundhouse.R
#' Roundhouse
#'
#' Generate a random Chuck Norris fact with a roundhouse kick.
#'
#' @param text Character string specifying the text to display with the
#' roundhouse (preferably, a Chuck Norris fact). If \code{NULL}, a fact is
#' selected at random from the ICNDb.
#'
#' @param width Integer specifying the number of characters per line.
#'
#' @param color Character string specifying the color of the text.
#'
#' @param fps Integer specifying the frames per second of the roundhouse kick.
#'
#' @param loop Integer specifying the number of times the roundhouse kick should
#' be repeated. Though a single roundhouse kick is always sufficient, the
#' default is \code{0} which results in an infinite number of roundhouse kicks.
#'
#' @param ... Additional optional arguments to be passed onto
#' \code{\link[magick]{image_annotate}}.
#'
#' @export
#'
#' @examples
#' roundhouse(width = 40, size = 15)
#'
#' roundhouse("Chuck Norris can’t test for equality because he has no equal.",
#'            width = 40, size = 25, fps = 5)
roundhouse <- function(text = NULL, width = 15, color = "white", fps = NULL,
                       loop = 0, ...) {
  if (is.null(text)) {
    text <- gsub(
      pattern = "&quot;",
      replacement = "'",
      x = random_fact()[["content"]][["value"]][[1L]][["joke"]]
    )
  }
  text <- paste0(strwrap(text, width = width), collapse = "\n")
  gif <- magick::image_read(
    path = "https://media.giphy.com/media/l1J3nY7N7LBrBobVm/giphy.gif"
  )
  if (is.null(fps)) {
    magick::image_annotate(gif, text = text, color = color, ...)
  } else {
    magick::image_animate(
      image = magick::image_annotate(gif, text = text, color = color, ...),
      fps = fps,
      dispose = "none",
      loop = loop
    )
  }
}


R/random_fact.R
#' Random Chuck Norris facts
#'
#' Query the internet Chuck Norris database (ICNDb) for a random Chuck Norris
#' fact.
#'
#' @param n Integer specifying the number of random facts to return. Default is
#' \code{1}.
#'
#' @param include Character string specifying the category of the joke. Default
#' is \code{NULL} meaning use all available categories in the query. The
#' available categories can be viewed using \code{fetch_categories()}.
#' (Currently ignored.)
#'
#' @param exclude Character string specifying the categories to exclude.
#' Default is \code{NULL} meaning use all available categories in the query.
#' The available categories can be viewed using \code{fetch_categories()}.
#' (Currently ignored.)
#'
#' @param sound Logical indicating whether or not "Chuck Norris" should be
#' shouted when generating random facts. Default is \code{FALSE}.
#'
#' @param sound_repeat Integer specifying the number of times the sound should
#' be repeated. Default is 1. (Only matters if \code{sound = TRUE}.)
#'
#' @return An object of class \code{"random_fact"}; essentially a character
#' vector.
#'
#' @rdname random_fact
#'
#' @export
#'
#' @examples
#' # Generate a random fact
#' random_fact()
#'
#' # Generate multiple random facts
#' random_fact(3, sound = TRUE)
random_fact <- function(n = 1L, include = NULL, exclude = NULL, sound = FALSE,
                        sound_repeat = 1) {

  # Play awesome sound (If requested)
  if (sound) {
    if (!requireNamespace("magick", quietly = TRUE)) {
      stop("The tuneR package is required when setting `sound = TRUE` for ",
           "additional awesomeness.", call. = FALSE)
    }
    mp3_file <- system.file("sounds", "chuck-norris.mp3",
                            package = "roundhouse")
    for (i in seq_len(sound_repeat)) {
      if (Sys.info()["sysname"] == "Darwin") {
        tuneR::play(tuneR::readMP3(mp3_file), player = "/usr/bin/afplay")
      } else {
        tuneR::play(tuneR::readMP3(mp3_file))
      }
    }
  }

  # Return random fact(s)
  roundhouse::facts[["fact"]][sample(nrow(roundhouse::facts), size = n,
                                     replace = FALSE)]

}


R/suggest_new_fact.R
#' Suggest new Chuck Norris facts
#'
#' Have a new Chuck Norris fact you want added to the database? Use this
#' function to open a new issue on roundhouse's GitHub page. If it's good, I'll
#' be sure to include it in the next release!
#'
#' @param fact Character string specifying a new Chuck Norris fact.
#'
#' @param user Character string specifying a valid GitHub username.
#'
#' @param password Character string specifying a valid password associated with
#' the above GitHub username.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' request_new_fact}
suggest_new_fact <- function(fact, user, password) {
  if (!requireNamespace("httr", quietly = TRUE)) {
    stop("The httr package is required for requesting new facts.",
         call. = FALSE)
  }
  r <- httr::POST(
    url = "https://api.github.com/repos/bgreenwell/roundhouse/issues",
    body = list(
      "title" = "New Chuck Norris Fact",
      "body" = fact
      # "labels" = "enhancement"
    ),
    encode = "json",
    httr::authenticate(user = user, password = password)
  )
  if (r$status_code == 201) {
    message("Request successfully submitted: ", r$headers$location)
  } else {
    message("Request was unsuccessful. Status code: ", r$status_code)
  }
}


tools/roundhouse-logo.R
# Load required packages
library(ggimage)
library(ggplot2)
library(grid)
library(png)

# Load vip image
img <- rasterGrob(readPNG("tools/chuck-norris.png"), interpolate = TRUE,
                  width = 0.5)

# Hexagon data
hex <- data.frame(x = 1.35 * 1 * c(-sqrt(3) / 2, 0, rep(sqrt(3) / 2, 2), 0,
                                   rep(-sqrt(3) / 2, 2)),
                  y = 1.35 * 1 * c(0.5, 1, 0.5, -0.5, -1, -0.5, 0.5))

# Hexagon logo
g <- ggplot() +
  geom_polygon(data = hex, aes(x, y), color = "black",
               fill = 'white', size = 3) +
  annotation_custom(img, xmin = -Inf, xmax = Inf, ymin = -0.5, ymax = 1) +
  annotate(geom = "text", x = 0, y = 0.6, color = "black", size = 7,
           label = "roundhouse") +
  coord_equal(xlim = range(hex$x), ylim = range(hex$y)) +
  scale_x_continuous(expand = c(0.04, 0)) +
  scale_y_reverse(expand = c(0.04, 0)) +
  theme_void() +
  theme_transparent() +
  theme(axis.ticks.length = unit(0, "mm"))
print(g)

png("tools/roundhouse-logo.png", width = 181, height = 209,
    bg = "transparent", type = "quartz")
print(g)
dev.off()

svg("tools/roundhouse-logo.svg", width = 181 / 72, height = 209 / 72,
    bg = "transparent")
print(g)
dev.off()


R/punch.R
#' Roundhouse punch
#'
#' Generate a roundhouse punch from Chuck Norris...with a random fact (or user-
#' supplied text).
#'
#' @param text Character string specifying the text to display with the
#' roundhouse (preferably, a Chuck Norris fact). If \code{NULL}, a fact is
#' selected at random from the ICNDb.
#'
#' @param width Integer specifying the number of characters per line. Default is
#' 35
#'
#' @param size Integer specifying the font size in pixels. Default is 10.
#'
#' @param color Character string specifying the color of the text.
#'
#' @param fps Integer specifying the frames per second of the roundhouse punch.
#' (Must be a factor of 100.)
#'
#' @param loop Integer specifying the number of times the roundhouse punch
#' should be repeated. Though a single roundhouse punch is always sufficient,
#' the default is \code{0} which results in an infinite number of roundhouse
#' punches.
#'
#' @param sound Logical indicating whether or not "Chuck Norris" should be
#' shouted when generating random facts. Default is \code{FALSE}.
#'
#' @param sound_repeat Integer specifying the number of times the sound should
#' be repeated. Default is 1. (Only matters if \code{sound = TRUE}.)
#'
#' @param sound_delay Integer specifying the delay (in seconds) to wait before
#' shouting "Chuck Norris". Default is 0. (Only matters if \code{sound = TRUE}.)
#'
#' @param ... Additional optional arguments to be passed onto
#' \code{\link[magick]{image_annotate}}.
#'
#' @export
#'
#' @examples
#' \dontrun{
#' # Random fact
#' punch(width = 40, size = 15)
#'
#' # User-supplied text
#' punch("Chuck Norris can’t test for equality because he has no equal.",
#'      type = 2, width = 40, size = 25, fps = 5)
#' }
punch <- function(text = NULL, width = 35, size = 20, color = "white",
                  fps = NULL, loop = 0, sound = FALSE, sound_repeat = 1,
                  sound_delay = 0, ...) {
  if (!requireNamespace("magick", quietly = TRUE)) {
    stop("The magick package is required for roundhouse punches.",
         call. = FALSE)
  }
  if (is.null(text)) {
    text <- random_fact(1L)
  }
  url <- "https://media.giphy.com/media/Fs8pXeLXuKUGA/giphy.gif"
  text <- paste0(strwrap(text, width = width), collapse = "\n")
  gif <- magick::image_read(path = url)
  res <- if (is.null(fps)) {
    magick::image_annotate(
      image = gif, text = text, color = color, size = size, ...
    )
  } else {
    magick::image_animate(
      image = magick::image_annotate(
        image = gif, text = text, color = color, size = size, ...
      ),
      fps = fps,
      dispose = "none",
      loop = loop
    )
  }

  # Display gif
  print(res)

  # Play awesome sound(s)
  if (sound) {
    Sys.sleep(sound_delay)
    mp3_file <- system.file("sounds", "chuck-norris.mp3",
                            package = "roundhouse")
    for (i in seq_len(sound_repeat)) {
      if (Sys.info()["sysname"] == "Darwin") {
        tuneR::play(tuneR::readMP3(mp3_file), player = "/usr/bin/afplay")
      } else {
        tuneR::play(tuneR::readMP3(mp3_file))
      }
    }
  }

  # Print GIF
  invisible(res)

}
